test/Microsoft.Azure.WebJobs.Host.EndToEndTests/ApplicationInsights/ApplicationInsightsEndToEndTests.cs

```diff
@@ -447,6 +455,236 @@ public async Task ApplicationInsights_HttpRequestTrackingByWebJobs(string testNa
+                {
+                    ["Links"] = new[] { new Activity("event hub process").SetParentId(parentId) }
+                }))
+                {
+                    MethodInfo methodInfo = GetType().GetMethod(testName, BindingFlags.Public | BindingFlags.Static);
+                    await host.GetJobHost().CallAsync(methodInfo, new { input = "function input" });
+                }
+
+                await host.StopAsync();
+
+                // Validate the request
+                RequestTelemetry functionRequest = _channel.Telemetries.OfType<RequestTelemetry>().SingleOrDefault();
+
+                 ValidateRequest(
+                    functionRequest,
+                    testName,
+                    testName,
+                    null,
+                    null,
+                    true,
+                    "0",
+                    traceId,
+                    $"|{traceId}.{parentSpanId}.");
+                Assert.Equal(sampledIn ? SamplingDecision.SampledIn : SamplingDecision.None, functionRequest.ProactiveSamplingDecision);                    
+
+                // Make sure operation ids match
+                var traces = _channel.Telemetries.OfType<TraceTelemetry>().Where(t => t.Context.Operation.Id == functionRequest.Context.Operation.Id);
+                Assert.Equal(4, traces.Count());
+                Assert.False(functionRequest.Properties.ContainsKey("_MS.links"));
+                Assert.False(functionRequest.Properties.ContainsKey("Links"));
+            }
+        }
+
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public async Task ApplicationInsights_EventHubTrackingByWebJobs_MultipleLinks(bool anySampledIn)
+        {
+            string testName = nameof(TestApplicationInsightsInformation);
+
+            using (IHost host = ConfigureHost())
+            {
+                Startup.Host = host;
+                await host.StartAsync();
+
+                var loggerProvider = host.Services.GetServices<ILoggerProvider>().OfType<ApplicationInsightsLoggerProvider>().Single();
+                var logger = loggerProvider.CreateLogger(LogCategories.Results);
+
+                var links = new Activity[2];
+                links[0] = new Activity("link0").SetParentId(
+                        ActivityTraceId.CreateRandom(), 
+                        ActivitySpanId.CreateRandom(), 
+                        anySampledIn ? ActivityTraceFlags.Recorded : ActivityTraceFlags.None);
+                links[1] = new Activity("link1").SetParentId(
+                    ActivityTraceId.CreateRandom(), 
+                    ActivitySpanId.CreateRandom(),
+                    ActivityTraceFlags.None);
+
+                // simulate functions behavior to set request on the scope
+                using (var _ = logger.BeginScope(new Dictionary<string, object>
+                {
+                    ["Links"] = links
```

Comment:
So these can be two, separate Activities? Is this how you deal with batches?

Reply:
so each message potentially carries a unique context. Each of the context in the message we call a 'link' and represent with Activity. This activity never starts (i.e. never becomes current)

When we track event hub trigger (a batch processing) we create a brand new Activity and request telemtery out of it.
We pass all these links to this new context. There are a few ways:
- from AppInsights point of view, this is done via custom _MS.links property on request telemetry - this is what we do here.
- links are also first-class citizen in the OpenTelemtery - OSS effort we heavily participate in and moving towards.
- we are also working with .NET folks to support links on Activity level

So the new distributed context is generated for batch, but knows about linked messages. UX will show them (on the Gantt chart first and eventually App-Map) and users will also be able to query them (this is advanced experience). UX experiences are in development now.